File: scripts/extract_frames.m
#import <Foundation/Foundation.h>
#import <AVFoundation/AVFoundation.h>
#import <CoreGraphics/CoreGraphics.h>
#import <ImageIO/ImageIO.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *videoPath = @"/Users/coi/Desktop/Renew/Fix/VideoMP4.MP4";
        NSString *outDir = @"/Users/coi/Desktop/Renew/Fix/frames";

        [[NSFileManager defaultManager] createDirectoryAtPath:outDir withIntermediateDirectories:YES attributes:nil error:nil];

        NSURL *url = [NSURL fileURLWithPath:videoPath];
        AVURLAsset *asset = [AVURLAsset URLAssetWithURL:url options:nil];

        Float64 durationSeconds = CMTimeGetSeconds(asset.duration);
        printf("Video duration: %.2f seconds\n", durationSeconds);

        AVAssetImageGenerator *generator = [[AVAssetImageGenerator alloc] initWithAsset:asset];
        generator.appliesPreferredTrackTransform = YES;
        generator.requestedTimeToleranceBefore = kCMTimeZero;
        generator.requestedTimeToleranceAfter = kCMTimeZero;

        int frameCount = 15;
        for (int i = 0; i < frameCount; i++) {
            Float64 t = (durationSeconds > 0) ? (durationSeconds * (double)i / (double)(frameCount - 1)) : 0;
            CMTime time = CMTimeMakeWithSeconds(t, 600);
            NSError *err = nil;
            CGImageRef cgImage = [generator copyCGImageAtTime:time actualTime:NULL error:&err];
            if (cgImage) {
                NSString *outPath = [NSString stringWithFormat:@"%@/frame_%02d_at_%ds.jpg", outDir, i, (int)t];
                NSURL *outUrl = [NSURL fileURLWithPath:outPath];
                CGImageDestinationRef dest = CGImageDestinationCreateWithURL((__bridge CFURLRef)outUrl, (CFStringRef)@"public.jpeg", 1, NULL);
                if (dest) {
                    NSDictionary *options = @{(__bridge NSString *)kCGImageDestinationLossyCompressionQuality: @(0.85)};
                    CGImageDestinationAddImage(dest, cgImage, (__bridge CFDictionaryRef)options);
                    CGImageDestinationFinalize(dest);
                    CFRelease(dest);
                    printf("Saved frame %d at %ds: %s\n", i, (int)t, [outPath UTF8String]);
                }
                CGImageRelease(cgImage);
            } else {
                printf("Error at %.1fs: %s\n", t, err ? [[err description] UTF8String] : "unknown");
            }
        }
    }
    return 0;
}

File: ds/scripts/extract_frames.swift
import Foundation
import AVFoundation
import CoreGraphics
import ImageIO

let videoPath = "/Users/coi/Desktop/Renew/Fix/VideoMP4.MP4"
let outDir = "/Users/coi/Desktop/Renew/Fix/frames"

let fileManager = FileManager.default
try? fileManager.createDirectory(atPath: outDir, withIntermediateDirectories: true, attributes: nil)

let url = URL(fileURLWithPath: videoPath)
let asset = AVURLAsset(url: url)

let durationSeconds = CMTimeGetSeconds(asset.duration)
print("Video duration: \(durationSeconds) seconds")

let generator = AVAssetImageGenerator(asset: asset)
generator.appliesPreferredTrackTransform = true
generator.requestedTimeToleranceBefore = .zero
generator.requestedTimeToleranceAfter = .zero

// Extract up to 15 key frames evenly spaced across the video
let frameCount = 15
for i in 0..<frameCount {
    let t = durationSeconds * Double(i) / Double(frameCount - 1)
    let time = CMTime(seconds: t, preferredTimescale: 600)
    do {
        let cgImage = try generator.copyCGImage(at: time, actualTime: nil)
        let outUrl = URL(fileURLWithPath: "\(outDir)/frame_\(String(format: "%02d", i))_at_\(Int(t))s.jpg")
        if let destination = CGImageDestinationCreateWithURL(outUrl as CFURL, "public.jpeg" as CFString, 1, nil) {
            let options: [CFString: Any] = [kCGImageDestinationLossyCompressionQuality: 0.85]
            CGImageDestinationAddImage(destination, cgImage, options as CFDictionary)
            CGImageDestinationFinalize(destination)
            print("Saved frame \(i) at \(Int(t))s: \(outUrl.lastPathComponent)")
        }
    } catch {
        print("Failed to extract frame at \(t)s: \(error)")
    }
}
